dev-java/jcommander: add 1.82

vaukai opened the pull request:
@Flowdalic 

This needs your attention since it involves a [java-util-2-runner-args.patch](https://github.com/gentoo/gentoo/pull/21319#issuecomment-878128343).  Without that patch `testng` won't tell what's wrong.
However I do not include this patch here as it's @fordfrog's work.

Furthermore I do not understand the remaining failure output.   
Which files are missing?
```
===============================================
Command line suite
Total tests run: 258, Failures: 2, Skips: 0
===============================================

java.io.IOException: No such file or directory
        at java.base/java.io.UnixFileSystem.createFileExclusively(Native Method)
        at java.base/java.io.File.createTempFile(File.java:2170)
        at java.base/java.io.File.createTempFile(File.java:2216)
        at org.testng.reporters.FileStringBuffer.flushToFile(FileStringBuffer.java:111)
        at org.testng.reporters.FileStringBuffer.append(FileStringBuffer.java:51)
        at org.testng.reporters.FileStringBuffer.append(FileStringBuffer.java:28)
        at org.testng.reporters.XMLUtils.xmlClose(XMLUtils.java:115)
        at org.testng.reporters.XMLStringBuffer.pop(XMLStringBuffer.java:182)
        at org.testng.reporters.jq.NavigatorPanel.generateMethodList(NavigatorPanel.java:240)
        at org.testng.reporters.jq.NavigatorPanel.generateResult(NavigatorPanel.java:115)
        at org.testng.reporters.jq.NavigatorPanel.generate(NavigatorPanel.java:70)
        at org.testng.reporters.jq.Main.generateReport(Main.java:51)
        at org.testng.TestNG.generateReports(TestNG.java:1135)
        at org.testng.TestNG.run(TestNG.java:1081)
        at org.testng.TestNG.privateMain(TestNG.java:1385)
        at org.testng.TestNG.main(TestNG.java:1354)
[TestNG] Reporter org.testng.reporters.jq.Main@6979e8cb failed
java.lang.NullPointerException
        at java.base/java.io.FileOutputStream.<init>(FileOutputStream.java:226)
        at java.base/java.io.FileWriter.<init>(FileWriter.java:113)
        at org.testng.reporters.FileStringBuffer.flushToFile(FileStringBuffer.java:120)
        at org.testng.reporters.FileStringBuffer.append(FileStringBuffer.java:51)
        at org.testng.reporters.FileStringBuffer.append(FileStringBuffer.java:28)
        at org.testng.reporters.XMLUtils.xmlClose(XMLUtils.java:115)
        at org.testng.reporters.XMLStringBuffer.pop(XMLStringBuffer.java:182)
        at org.testng.reporters.jq.NavigatorPanel.generateMethodList(NavigatorPanel.java:240)
        at org.testng.reporters.jq.NavigatorPanel.generateResult(NavigatorPanel.java:115)
        at org.testng.reporters.jq.NavigatorPanel.generate(NavigatorPanel.java:70)
        at org.testng.reporters.jq.Main.generateReport(Main.java:51)
        at org.testng.TestNG.generateReports(TestNG.java:1135)
        at org.testng.TestNG.run(TestNG.java:1081)
        at org.testng.TestNG.privateMain(TestNG.java:1385)
        at org.testng.TestNG.main(TestNG.java:1354
```

Flowdalic commented on dev-java/jcommander/jcommander-1.82.ebuild:
```diff
@@ -0,0 +1,56 @@
+src_test() {
+	# because module java.base does not export sun.reflect.annotation to unnamed module @0x42bb2aee
+	local vm_version="$(java-config -g PROVIDES_VERSION)"
+	if [[ "${vm_version}" -ge "17" ]] ; then
```
No need to quote the 17

vaukai replied:
Right.  We've already that case somewhere.
Adjusted to using the same construction as in `jackson-dataformat-xml`.

Flowdalic commented:
It looks like two unit tests failing, and then the report generation fails also with a prime example of an exception message that is lacking contextual information. But from the stacktrace, I'd guess that it is because the JVM is unable to determine a tempdir. You probably want to add `-Djava.io.tmpdir=\"${T}\"` JVM argument when running the tests.

In the meantime, we should also focus on getting @fordfrog's eclass patch into Gentoo. @fordfrog fancy to create a PR for this?

vaukai commented:
> It looks like two unit tests failing, and then the report generation fails also with a prime example of an exception message that is lacking contextual information. But from the stacktrace, I'd guess that it is because the JVM is unable to determine a tempdir. You probably want to add `-Djava.io.tmpdir=\"${T}\"` JVM argument when running the tests.
> 
Just added this line
```
JAVA_TEST_EXTRA_ARGS="-Djava.io.tmpdir=${T}"
```
But it won't tell us what was wrong - it simply solves it.

> In the meantime, we should also focus on getting @fordfrog's eclass patch into Gentoo. @fordfrog fancy to create a PR for this?

In the second commit I added the eclass patch for review. Guess it needs @fordfrog's signed-off.

Flowdalic commented on dev-java/jcommander/jcommander-1.82.ebuild:
```diff
@@ -0,0 +1,56 @@
+"
+RDEPEND=">=virtual/jre-1.8:*"
+
+S="${WORKDIR}/${P}"
```
That's the default value for `S`, isn't it? If so, then this can be omitted.

vaukai replied:
java-pkg-simple.eclass defaults to `S="${WORKDIR}"`.

Flowdalic commented on dev-java/jcommander/jcommander-1.82.ebuild:
```diff
@@ -0,0 +1,56 @@
+SLOT="0"
+KEYWORDS="~amd64 ~arm ~arm64 ~ppc64 ~x86"
+
+DEPEND=">=virtual/jdk-1.8:*
```
Personally, I would put multiple dependencies on their own lines. So

```
DEPEND="
    test? ( dev-java/jackson-annotations:2 )
    >=virtual/jdk-1.8:*
```

Flowdalic commented on dev-java/jcommander/jcommander-1.82.ebuild:
```diff
@@ -0,0 +1,60 @@
+# Needs "java-util-2-runner-args.patch" from https://github.com/gentoo/gentoo/pull/21319#issuecomment-878128343
+JAVA_TEST_RUNNER_EXTRA_ARGS=( -verbose 3 -usedefaultlisteners true )
+
+JAVA_TEST_EXTRA_ARGS="-Djava.io.tmpdir=${T}"
```
If it also works, then this should probably be

```suggestion
JAVA_TEST_EXTRA_ARGS="-Djava.io.tmpdir=\"${T}\""
```

in case `T` contains a white space. I am not absolutely sure when it comes to JVM `-D` option parsing, so I don't know if it does the right thing, but if it doesn't break anything, then this is probably better.

Flowdalic commented:
It appears the eclass change is not strictly required for this to work. It only improves the output situation for `src_test`, right? If so, then please drop the commit from this PR. I'll create an extra PR for the eclass change and submit it to @gentoo-dev.

vaukai commented:
> It appears the eclass change is not strictly required for this to work. It only improves the output situation for `src_test`, right? If so, then please drop the commit from this PR. I'll create an extra PR for the eclass change and submit it to @gentoo-dev.

Yes that's correct. Commit dropped.
